### PR TITLE
Put completion in 'suggestion' mode when typing a lambda to a method that could be an extension or instance method.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/28586

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@sharwell @dotnet/roslyn-ide @jcouv can i get eyes on this?  Thanks!

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs [hunk 1]
```diff
@@ -843,6 +843,230 @@ void A()
...
         }
 
+        [WorkItem(28586, "https://github.com/dotnet/roslyn/issues/28586")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task WithExtensionAndInstanceMethod1()
+        {
+            var markup = @"
+using System;
+
+public sealed class Foo
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs]: >Foo [](start = 20, length = 3)

nit: policheck ;-)

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs [hunk 2]
```diff
@@ -843,6 +843,230 @@ void A()
...
+    public static void Bar(this Foo foo, Action<int> action)
+    {
+    }
+}
+
+public static class Repro
+{
+    public static void ReproMethod(Foo foo)
+    {
+        foo.Bar(($$
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs]: >( [](start = 16, length = 1)

From a quick look at `CSharpSuggestionModeCompletionProvider.IsLambdaExpression`, I would think this should have worked already (it handles a [case for open paren](https://github.com/dotnet/roslyn/blob/03548b467e6e8844db676afd6dcd616c173265b4/src/Features/CSharp/Portable/Completion/SuggestionMode/CSharpSuggestionModeCompletionProvider.cs#L123)).

Yet the bug report indicates that didn't work. What part of your change fixed this?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs]: That checks for specifically: ``` if (token.Parent != null && token.Parent is ParameterListSyntax)```

but the above is not a parameter list.  it's a ParenthesizedExpression.

Also, a bunch of these tests were purely to try to test *around* this feature to make sure lots of cases work properly.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs]: The main part of interest is: 

https://github.com/dotnet/roslyn/blob/03548b467e6e8844db676afd6dcd616c173265b4/src/Features/CSharp/Portable/Completion/SuggestionMode/CSharpSuggestionModeCompletionProvider.cs#L182-L183

We basically dont' know if this is a lambda or not (syntactically) so we use the semantic type-inference service to try to figure out.  However, that service didn't do well in this specific overload resolution case.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs [hunk 3]
```diff
@@ -843,6 +843,230 @@ void A()
...
+public static class Repro
+{
+    public static void ReproMethod(Foo foo)
+    {
+        foo.Bar(async (a$$
+    }
+}
+";
+            await VerifyBuilderAsync(markup);
+        }
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs]: >} [](start = 8, length = 1)

Not related to this PR: I didn't see any test with a parameter that's not a `Func` or `Action` (ie. no suggestion mode).

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs]: Done.  Sigh :)

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide can i get more eyes?  Thanks!

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide @jasonmalinowski  can i get more eyes? Thanks!

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide @dpoeschl can i get some eyes on this pretty simple PR?  Thanks! :)

**Comment by CyrusNajmabadi** [comment]
test windows_release_vs-integration_prtest

**Comment by CyrusNajmabadi** [comment]
test windows_debug_vs-integration_prtest

**Pull request merged**